# Aviation Accidents Analysis

You are part of a consulting firm that is tasked to do an analysis of commercial and passenger jet airline safety. The client (an airline/airplane insurer) is interested in knowing what types of aircraft (makes/models) exhibit low rates of total destruction and low likelihood of fatal or serious passenger injuries in the event of an accident. They are also interested in any general variables/conditions that might be at play. Your analysis will be based off of aviation accident data accumulated from the years 1948-2023. 

Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e. from 1983 onwards). They would also like separate recommendations for small aircraft vs. larger passenger models. **In addition, make sure that claims that you make are statistically robust and that you have enough samples when making comparisons between groups.**


In this summative assessment you will demonstrate your ability to:
- **Use Pandas to load, inspect, and clean the dataset appropriately.**
- **Transform relevant columns to create measures that address the problem at hand.**
- conduct EDA: visualization and statistical measures to systematically understand the structure of the data
- recommend a set of airplanes and makes conforming to the client's request and identify at least *two* factors contributing to airplane safety. You must provide supporting evidence (visuals, summary statistics, tables) for each claim you make.

### Make relevant library imports

In [355]:
import pandas as pd
import numpy as np

## Data Loading and Inspection

### Load in data from the relevant directory and inspect the dataframe.
- inspect NaNs, datatypes, and summary statistics

In [356]:
df = pd.read_csv('C:/Users/user/Aviation_Assessment/Summative_Assessment_Lab/data/AviationData.csv', encoding='latin1')
df.head()

FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/user/Aviation_Assessment/Summative_Assessment_Lab/data/AviationData.csv'

In [ ]:
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [ ]:
df.info

<bound method DataFrame.info of              Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude  Longitude Airport.Code  \
0      MOOSE CREEK, ID  Uni

The dataset has 88,889 rows and 31 columns.

In [ ]:
df.dtypes

Event.Id                   object
Investigation.Type         object
Accident.Number            object
Event.Date                 object
Location                   object
Country                    object
Latitude                   object
Longitude                  object
Airport.Code               object
Airport.Name               object
Injury.Severity            object
Aircraft.damage            object
Aircraft.Category          object
Registration.Number        object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
FAR.Description            object
Schedule                   object
Purpose.of.flight          object
Air.carrier                object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
Report.Status 

Most of the data here are strings. Some of them will be changed to the correct data type.

In [ ]:
df.duplicated().sum()

np.int64(0)

There were no duplicates in the dataset

In [ ]:
df.isnull().sum().sort_values(ascending=False)

Schedule                  76307
Air.carrier               72241
FAR.Description           56866
Aircraft.Category         56602
Longitude                 54516
Latitude                  54507
Airport.Code              38757
Airport.Name              36185
Broad.phase.of.flight     27165
Publication.Date          13771
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Fatal.Injuries      11401
Engine.Type                7096
Report.Status              6384
Purpose.of.flight          6192
Number.of.Engines          6084
Total.Uninjured            5912
Weather.Condition          4492
Aircraft.damage            3194
Registration.Number        1382
Injury.Severity            1000
Country                     226
Amateur.Built               102
Model                        92
Make                         63
Location                     52
Investigation.Type            0
Event.Date                    0
Accident.Number               0
Event.Id                      0
dtype: i

There were quite a large number of missing values in some of the columns in the dataset. Handling of the missing values will mainly be focused on the columns that satisfy the client's needs from the analysis.

I decided to drop the rows with the missing values in Make, Model and Amateur bulit columns because they make up less than 1% of the dataset and also, without values in those columns, those rows are not valuable anymore.

In [ ]:
df = df.dropna(subset=['Make', 'Model', 'Amateur.Built'])

In [ ]:
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

## Data Cleaning

### Filtering aircrafts and events

We want to filter the dataset to include aircraft that the client is interested in an analysis of:
- inspect relevant columns
- figure out any reasonable imputations
- filter the dataset

The code below changes the Event date data type to date.

In [ ]:
df['Event.Date'] = pd.to_datetime(df['Event.Date'])

C:\Users\user\AppData\Local\Temp\ipykernel_5276\1540095821.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Event.Date'] = pd.to_datetime(df['Event.Date'])


The code below filters the data by the event date specified by the client, that is from 1983 to 2023.

In [ ]:
df = df[df['Event.Date'] >= '1983-01-01']

The code below filters the dataset by airplanes since the client is only interested in airplanes.

In [ ]:
df = df[df['Aircraft.Category'] == 'Airplane']

The code below filters the data by the Amateur Built, since the client wants airplanes that were proffesionally built only.

In [ ]:
df = df[df['Amateur.Built'] == 'No']

The code below filters the data by the investigation type, since the client is already interested in accidents and not incidents.

In [ ]:
df = df[df['Investigation.Type'] == 'Accident']

### Cleaning and constructing Key Measurables

Injuries and robustness to destruction are a key interest point for the client. Clean and impute relevant columns and then create derived fields that best quantifies what the client wishes to track. **Use commenting or markdown to explain any cleaning assumptions as well as any derived columns you create.**

**Construct metric for fatal/serious injuries**

*Hint:* Estimate the total number of passengers on each flight. The likelihood of serious / fatal injury can be estimated as a fraction from this.

This is the total number of passengers on each flight

I chose to assume that if no injuries were recorded, then it means there were no injuries. So i filled all the missing values with 0.

In [ ]:
injury_cols = ['Total.Fatal.Injuries', 'Total.Serious.Injuries', 
               'Total.Minor.Injuries', 'Total.Uninjured']
df[injury_cols] = df[injury_cols].fillna(0)

The code below gives the total number of passengers per flight.

In [ ]:
df['Total.Passengers'] = df[injury_cols].sum(axis=1)

Below is the code to check the rate of Fatal/Seroius injuries.

In [ ]:
df['Fatal.Serious.Rate'] = (df['Total.Fatal.Injuries'] + df['Total.Serious.Injuries']) / df['Total.Passengers']

**Aircraft.Damage**
- identify and execute any cleaning tasks
- construct a derived column tracking whether an aircraft was destroyed or not.

The code below fills in all missing values in the aircraft damage column with unknown. This assumes that if the type of damage was not recorded, then it was unknown.

In [ ]:
df['Aircraft.damage'].fillna('Unknown')

4171       Destroyed
4285         Unknown
5960       Destroyed
6669         Unknown
6806     Substantial
            ...     
88869    Substantial
88873    Substantial
88876    Substantial
88877    Substantial
88886    Substantial
Name: Aircraft.damage, Length: 19923, dtype: object

The code creates a column called Destroyed.Aircrafts, which with 1, it means that aircraft damage was totally Destroyed, which is what the clients wants to know. If it shows 0, it means for those aircrafts the damage was either Unknown, Subsatantial or Minor.

In [ ]:
df['Destroyed.Aircrafts'] = (df['Aircraft.damage'] == 'Destroyed').astype(int)

### Investigate the *Make* column
- Identify cleaning tasks here
- List cleaning tasks clearly in markdown
- Execute the cleaning tasks
- For your analysis, keep Makes with a reasonable number (you can put the threshold at 50 though lower could work as well)

The code below makes the necessary replacements in the file.

In [ ]:
make_replacements = {
    'Dehavilland': 'De Havilland',
    'Air Tractor Inc': 'Air Tractor',
    'Aviat Aircraft Inc': 'Aviat',
    'Ayres Corporation': 'Ayres',
    'Cirrus Design Corp': 'Cirrus',
    'Diamond Aircraft Ind Inc': 'Diamond',
    'Mooney Aircraft Corp.': 'Mooney',
}
df['Make'] = df['Make'].replace(make_replacements)

In [ ]:
df['Make'] = df['Make'].str.title().str.strip()

The code above ensures there are no unwanted whitespaces and the casings are the same.

In [ ]:
make_counts = df['Make'].value_counts()
valid_makes = make_counts[make_counts >= 30].index
df = df[df['Make'].isin(valid_makes)]


I have chosen 30 as the threshold as 30.

### Inspect Model column
- Get rid of any NaNs.
- Inspect the column and counts for each model/make. Are model labels unique to each make?
- If not, create a derived column that is a unique identifier for a given plane type.

In [ ]:
df['Model'].isna().sum()

np.int64(0)

In [ ]:
df['Model'] = df['Model'].str.upper().str.strip()

The code below confirms that not every model label is unique to the make.

In [ ]:
df.groupby('Model')['Make'].nunique().sort_values(ascending=False)

Model
S2R                 5
S 2R                4
S-2R                4
112                 3
8KCAB               3
                   ..
A380                1
A36TC               1
A36 - UNDESIGNAT    1
A36 - TC            1
YPF-7               1
Name: Make, Length: 1959, dtype: int64

The above code creates a new column 'Make_Model' which becomes the unique identifier for a given plane type. It combines the Make and Model into one name.

In [ ]:
df['Make_Model'] = df['Make'] + ' ' + df['Model']

### Cleaning other columns
- there are other columns containing data that might be related to the outcome of an accident. We list a few here:
- Engine.Type
- Weather.Condition
- Number.of.Engines
- Purpose.of.flight
- Broad.phase.of.flight

Inspect and identify potential cleaning tasks in each of the above columns. Execute those cleaning tasks. 

**Note**: You do not necessarily need to impute or drop NaNs here.

In [ ]:
df['Engine.Type'].value_counts(dropna=False)

Engine.Type
Reciprocating    13013
NaN               2520
Turbo Prop         933
Turbo Fan          402
Unknown             71
Turbo Jet           57
Turbo Shaft          9
UNK                  1
Name: count, dtype: int64

In [ ]:
df['Engine.Type'] = df['Engine.Type'].replace({'UNK': 'Unknown'})

Th unknown values in the engine type column have been standardized.

In [ ]:
df['Weather.Condition'].value_counts(dropna=False)

Weather.Condition
VMC    14369
NaN     1549
IMC      892
Unk      150
UNK       46
Name: count, dtype: int64

In [ ]:
df['Weather.Condition'] = df['Weather.Condition'].replace({'Unk': 'Unknown', 'UNK': 'Unknown'})

The unknown values in the weather condition column have been standardized.

In [ ]:
df['Number.of.Engines'].value_counts(dropna=False)

Number.of.Engines
1.0    13455
2.0     2026
NaN     1470
4.0       35
3.0       16
0.0        4
Name: count, dtype: int64

In [ ]:
df = df[df['Number.of.Engines'] != 0]

Records with 0 number of engines have been dropped, since they are not physically logical and they hold a small percentage of the data and dropping them does npt affect the analysis greatly.

In [ ]:
df['Purpose.of.flight'].value_counts(dropna=False)

Purpose.of.flight
Personal                     9982
Instructional                2457
NaN                          1986
Aerial Application            771
Business                      401
Positioning                   264
Unknown                       262
Skydiving                     156
Aerial Observation            149
Other Work Use                126
Banner Tow                     86
Flight Test                    74
Ferry                          74
Executive/corporate            61
Public Aircraft - Federal      30
Glider Tow                     29
Public Aircraft                25
Public Aircraft - State        21
Air Race show                  15
Firefighting                   13
Public Aircraft - Local        11
PUBS                            3
Air Race/show                   2
Air Drop                        2
ASHO                            2
Name: count, dtype: int64

In [ ]:
df['Purpose.of.flight'] = df['Purpose.of.flight'].replace({'Air Race/show': 'Air Race show', 'Public Aircraft - Federal': 'Public Aircraft',
    'Public Aircraft - State': 'Public Aircraft',
    'Public Aircraft - Local': 'Public Aircraft'})

In the purpose of flight column, the subcategories of Public aircraft have been merged into one, since the distinctions do not matter to the analysis.

In [ ]:
df['Broad.phase.of.flight'].value_counts(dropna=False)

Broad.phase.of.flight
NaN            14551
Landing         1119
Takeoff          425
Cruise           237
Approach         212
Maneuvering      130
Taxi              92
Go-around         82
Descent           59
Climb             48
Standing          35
Unknown           10
Other              2
Name: count, dtype: int64

The Broad phase of flight column does not need any cleaning.

### Column Removal
- inspect the dataframe and drop any columns that have too many NaNs

The columns below have been dropped because they contain a large number of missing values.

In [ ]:
df.drop(columns=[
    'Schedule',            
    'Air.carrier',         
    'FAR.Description',     
    'Longitude',           
    'Latitude',            
    'Airport.Code',        
    'Airport.Name',        
    'Publication.Date',    
], inplace=True)

In [ ]:
df.shape

(17002, 27)

After several cleaning steps the dataset has 17,002 rows and 27 columns, compared to 88,889 rows and 31 columns of the raw data.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17002 entries, 4171 to 88886
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Event.Id                17002 non-null  object        
 1   Investigation.Type      17002 non-null  object        
 2   Accident.Number         17002 non-null  object        
 3   Event.Date              17002 non-null  datetime64[ns]
 4   Location                16999 non-null  object        
 5   Country                 17001 non-null  object        
 6   Injury.Severity         16816 non-null  object        
 7   Aircraft.damage         16624 non-null  object        
 8   Aircraft.Category       17002 non-null  object        
 9   Registration.Number     16876 non-null  object        
 10  Make                    17002 non-null  object        
 11  Model                   17002 non-null  object        
 12  Amateur.Built           17002 non-null  object  

In [ ]:
df.isna().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                      3
Country                       1
Injury.Severity             186
Aircraft.damage             378
Aircraft.Category             0
Registration.Number         126
Make                          0
Model                         0
Amateur.Built                 0
Number.of.Engines          1470
Engine.Type                2520
Purpose.of.flight          1986
Total.Fatal.Injuries          0
Total.Serious.Injuries        0
Total.Minor.Injuries          0
Total.Uninjured               0
Weather.Condition          1549
Broad.phase.of.flight     14551
Report.Status              3015
Total.Passengers              0
Fatal.Serious.Rate          205
Destroyed.Aircrafts           0
Make_Model                    0
dtype: int64

### Save DataFrame to csv
- its generally useful to save data to file/server after its in a sufficiently cleaned or intermediate state
- the data can then be loaded directly in another notebook for further analysis
- this helps keep your notebooks and workflow readable, clean and modularized

In [ ]:
df.to_csv('AviationData.csv', index=False)

In [ ]:
check = pd.read_csv('AviationData.csv')
check.shape

(17002, 27)

In [ ]:
check.head

<bound method NDFrame.head of              Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001214X42331           Accident      ATL83FA140  1983-03-20   
1      20001214X42672           Accident      FTW83LA177  1983-04-02   
2      20001214X45188           Accident      NYC84LA028  1983-11-13   
3      20001214X45339           Accident      LAX84LA110  1983-12-22   
4      20001214X38957           Accident      ATL84LA120  1984-03-14   
...               ...                ...             ...         ...   
16997  20221212106444           Accident      ERA23LA085  2022-12-12   
16998  20221213106455           Accident      WPR23LA065  2022-12-13   
16999  20221215106463           Accident      ERA23LA090  2022-12-14   
17000  20221219106470           Accident      ERA23LA091  2022-12-16   
17001  20221227106497           Accident      WPR23LA075  2022-12-26   

                    Location        Country Injury.Severity Aircraft.damage  \
0             CROSSVILLE, 

The code above confirms if the saving worked.